File: output/sonarqube_analysis/amazonq_solutions/php/results/q45-dice-roll-simulation.php
<?php
/** https://leetcode.com/problems/dice-roll-simulation */
// Problem: Dice Roll Simulation

//Problem Description:
//A die simulator generates a random number from 1 to 6 for each roll. You introduced a constraint to the generator such that it cannot roll the number i more than rollMax[i] (1-indexed) consecutive times.
//Given an array of integers rollMax and an integer n, return the number of distinct sequences that can be obtained with exact n rolls. Since the answer may be too large, return it modulo 109 + 7.
//Two sequences are considered different if at least one element differs from each other.

// Constraints:
// 1 <= n <= 5000
// rollMax.length == 6
// 1 <= rollMax[i] <= 15

//Code Structure:

class Solution {

    /**
     * @param Integer $n
     * @param Integer[] $rollMax
     * @return Integer
     */
    function dieSimulator($n, $rollMax) {
        $mod = 1000000007;
        $dp = array_fill(0, $n + 1, array_fill(0, 6, array_fill(0, 16, 0)));
        $dp[0][0][0] = 1;

        for ($i = 1; $i <= $n; $i++) {
            for ($j = 0; $j < 6; $j++) {
                for ($k = 1; $k <= $rollMax[$j]; $k++) {
                    if ($k > $i) {
                        continue;
                    }
                    for ($l = 0; $l < 6; $l++) {
                        if ($l != $j) {
                            $dp[$i][$j][$k] = ($dp[$i][$j][$k] + $dp[$i - 1][$l][1]) % $mod;
                        } else {
                            $dp[$i][$j][$k] = ($dp[$i][$j][$k] + $dp[$i - 1][$l][$k - 1]) % $mod;
                        }
                    }
                }
            }
        }

        $result = 0;
        for ($i = 0; $i < 6; $i++) {
            $result = ($result + $dp[$n][$i][$rollMax[$i]]) % $mod;
        }

        return $result;
    }
}
    

File: output/sonarqube_analysis/amazonq_solutions/java/results/q45-dice-roll-simulation.java
/** https://leetcode.com/problems/dice-roll-simulation */
// Problem: Dice Roll Simulation

//Problem Description:
//A die simulator generates a random number from 1 to 6 for each roll. You introduced a constraint to the generator such that it cannot roll the number i more than rollMax[i] (1-indexed) consecutive times.
//Given an array of integers rollMax and an integer n, return the number of distinct sequences that can be obtained with exact n rolls. Since the answer may be too large, return it modulo 109 + 7.
//Two sequences are considered different if at least one element differs from each other.

// Constraints:
// 1 <= n <= 5000
// rollMax.length == 6
// 1 <= rollMax[i] <= 15

//Code Structure:

class Solution {
    public int dieSimulator(int n, int[] rollMax) {
        int MOD = 1000000007;
        int[][][] dp = new int[n + 1][6][16];
        
        for (int i = 0; i < 6; i++) {
            dp[1][i][1] = 1;
        }
        
        for (int i = 2; i <= n; i++) {
            for (int j = 0; j < 6; j++) {
                for (int k = 0; k < 6; k++) {
                    if (j == k) {
                        for (int c = 1; c < rollMax[j]; c++) {
                            dp[i][j][c + 1] = (dp[i][j][c + 1] + dp[i - 1][k][c]) % MOD;
                        }
                    } else {
                        for (int c = 1; c <= rollMax[k]; c++) {
                            dp[i][j][1] = (dp[i][j][1] + dp[i - 1][k][c]) % MOD;
                        }
                    }
                }
            }
        }
        
        int result = 0;
        for (int i = 0; i < 6; i++) {
            for (int j = 1; j <= rollMax[i]; j++) {
                result = (result + dp[n][i][j]) % MOD;
            }
        }
        return result;
    }
}
    